embedded-tests/src/test/java/org/apache/druid/testing/embedded/k8s/K3sClusterResource.java

```diff
@@ -0,0 +1,369 @@
+import org.apache.druid.testing.embedded.docker.DruidContainerResource;
+import org.apache.druid.testing.embedded.indexing.Resources;
+import org.apache.druid.testing.tools.ITRetryUtil;
+import org.testcontainers.k3s.K3sContainer;
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.MountableFile;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+
+public class K3sClusterResource extends TestcontainerResource<K3sContainer>
+{
+  private static final Logger log = new Logger(K3sClusterResource.class);
+
+  private static final String K3S_IMAGE_NAME = "rancher/k3s:v1.28.8-k3s1";
+
+  public static final String DRUID_NAMESPACE = "druid";
+  private static final String NAMESPACE_MANIFEST = "manifests/druid-namespace.yaml";
+
+  private static final String COMMON_CONFIG_MAP = "druid-common-props";
+  private static final String SERVICE_CONFIG_MAP = "druid-%s-props";
+
+  public static final long POD_READY_TIMEOUT_SECONDS = 300;
+
+  private final List<File> manifestFiles = new ArrayList<>();
+  private final List<K3sDruidService> services = new ArrayList<>();
+
+  private KubernetesClient client;
+  private String druidImageName;
+
+  private final Closer closer = Closer.create();
+
+  public K3sClusterResource()
+  {
+    // Add the namespace manifest
+    manifestFiles.add(Resources.getFileForResource(NAMESPACE_MANIFEST));
+  }
+
+  public K3sClusterResource addService(K3sDruidService service)
+  {
+    services.add(service);
+    return this;
+  }
+
+  public K3sClusterResource usingDruidImage(String druidImageName)
+  {
+    this.druidImageName = druidImageName;
+    return this;
+  }
+
+  /**
+   * Uses the Docker test image specified by the system property
+   * {@link DruidContainerResource#PROPERTY_TEST_IMAGE} for this container.
+   */
+  public K3sClusterResource usingTestImage()
+  {
+    final String imageName = System.getProperty(DruidContainerResource.PROPERTY_TEST_IMAGE);
+    InvalidInput.conditionalException(
+        imageName != null,
+        StringUtils.format(
+            "System property[%s] must be set while running Docker tests locally"
+            + " to specify which Druid image to use. Update your run configuration"
+            + " to include '-D%s=<your-test-image>'.",
+            DruidContainerResource.PROPERTY_TEST_IMAGE,
+            DruidContainerResource.PROPERTY_TEST_IMAGE
+        )
+    );
+    return usingDruidImage(imageName);
+  }
+
+  @Override
+  protected K3sContainer createContainer()
+  {
+    Objects.requireNonNull(druidImageName, "No Druid image specified");
```

Comment:
Nit: If this is a not null construct for container to be created, imo it makes more sense to pass it in the constructor itself rather than via `.usingDruidImage()`